### PR TITLE
CLI - Add command sb benchmark [list,list-parameters]

### PR DESCRIPTION
**Pull request by abuccts**
__Description__

Add command `sb benchmark list` and `sb benchmark list-parameters` to support listing all optional parameters for benchmarks.

<details>
<summary>Examples</summary>
<pre>
$ sb benchmark list -n [a-z]+-bw -o table
Result
--------
mem-bw
nccl-bw
rccl-bw
</pre>
<pre>
$ sb benchmark list-parameters -n mem-bw
=== mem-bw ===
optional arguments:
  --bin_dir str         Specify the directory of the benchmark binary.
  --duration int        The elapsed time of benchmark in seconds.
  --mem_type str [str ...]
                        Memory types to benchmark. E.g. htod dtoh dtod.
  --memory str          Memory argument for bandwidthtest. E.g. pinned unpinned.
  --run_count int       The run count of benchmark.
  --shmoo_mode          Enable shmoo mode for bandwidthtest.
default values:
{'bin_dir': None,
 'duration': 0,
 'mem_type': ['htod', 'dtoh'],
 'memory': 'pinned',
 'run_count': 1}
</pre>
</details>

__Major Revisions__
* Add `sb benchmark list` to list benchmarks matching given name.
* Add `sb benchmark list-parameters` to list parameters for benchmarks which match given name.

__Minor Revisions__
* Sort format help text for argparse.

### CODE REVIEW

File: superbench/cli/_commands.py [hunk 1]
```diff
@@ -23,6 +23,9 @@ def load_command_table(self, args):
             g.command('deploy', 'deploy_command_handler')
```
Comment (cp5555) [inline superbench/cli/_commands.py:23]: Please update the document as well.

File: docs/cli.md [hunk 2]
```diff
@@ -24,6 +24,68 @@ Welcome to the SB CLI!
...
+
+| Name          | Default | Description                           |
+|---------------|---------|---------------------------------------|
+| `--name` `-n` | `None`  | Benchmark name or regular expression. |
+
+#### Global arguments
+
+| Name          | Default | Description        |
+|---------------|---------|--------------------|
+| `--help` `-h` | N/A     | Show help message. |
```
Comment (cp5555) [inline docs/cli.md:44]: Why `help` here??

File: docs/cli.md [hunk 3]
```diff
@@ -24,6 +24,68 @@ Welcome to the SB CLI!
...
+
+List all benchmarks ending with "-bw":
+```bash title="SB CLI"
+sb benchmark list --name [a-z]+-bw
+```
+
+### `sb benchmark list-parameters`
+
+List parameters for benchmarks which match the name.
+```bash title="SB CLI"
```
Comment (cp5555) [inline docs/cli.md:61]: The bash title is strange.

File: docs/cli.md [hunk 4]
```diff
@@ -24,6 +24,68 @@ Welcome to the SB CLI!
 
 The following lists `sb` commands usages and examples:
 
+### `sb benchmark list`
+
+List benchmarks which match the name.
```
Comment (cp5555) [inline docs/cli.md:29]: If I know the name, why do we list it? What's the goal for this feature? I think you would like to list all benchmarks.

File: docs/cli.md [hunk 5]
```diff
@@ -24,6 +24,68 @@ Welcome to the SB CLI!
...
+List all benchmarks ending with "-bw":
+```bash title="SB CLI"
+sb benchmark list --name [a-z]+-bw
+```
+
+### `sb benchmark list-parameters`
+
+List parameters for benchmarks which match the name.
+```bash title="SB CLI"
+sb benchmark list-parameters [--name]
```
Comment (cp5555) [inline docs/cli.md:62]: If we have --help, does it should be sb benchmark list-parameters [--name] [--help]

File: tests/cli/test_sb.py [hunk 6]
```diff
@@ -83,6 +84,20 @@ def test_sb_run_nonexist_host_file(self):
...
+
+    def test_sb_benchmark_list_nonexist(self):
+        """Test sb benchmark list, give a non-exist benchmark name, should fail."""
+        result = self.cmd('sb benchmark list -n non-exist-name', expect_failure=True)
+        self.assertEqual(result.exit_code, 1)
+
+    def test_sb_benchmark_list_parameters(self):
+        """Test sb benchmark list-parameters."""
+        self.cmd('sb benchmark list-parameters', checks=[NoneCheck()])
+        self.cmd('sb benchmark list-parameters -n pytorch-[a-z]+', checks=[NoneCheck()])
```
Comment (cp5555) [inline tests/cli/test_sb.py:99]: Why is NoneCheck()?

File: superbench/cli/_help.py [hunk 7]
```diff
@@ -61,8 +61,31 @@
...
+    short-summary: Commands to manage benchmarks.
+"""
+
+helps['benchmark list'] = """
+    type: command
+    examples:
+        - name: list all benchmarks
+          text: {cli_name} benchmark list
+        - name: list all benchmarks ending with "-bw"
+          text: {cli_name} benchmark list --name [a-z]+-bw
```
Comment (cp5555) [inline superbench/cli/_help.py:75]: Why do we end with `-bw`

File: superbench/cli/_help.py [hunk 8]
```diff
@@ -61,8 +61,31 @@
...
+""".format(cli_name=CLI_NAME)
+
+helps['benchmark list-parameters'] = """
+    type: command
+    examples:
+        - name: list parameters for all benchmarks
+          text: {cli_name} benchmark list-parameters
+        - name: list parameters for all benchmarks which starts with "pytorch-"
```
Comment (cp5555) [inline superbench/cli/_help.py:83]: Also why is `pytorch-`?

File: superbench/benchmarks/base.py [hunk 9]
```diff
@@ -6,13 +6,25 @@
...
 
+class SortedMetavarTypeHelpFormatter(argparse.MetavarTypeHelpFormatter):
+    """Custom HelpFormatter class for argparse which sorts option strings."""
+    def add_arguments(self, actions):
+        """Sort option strings before original add_arguments.
+
+        Args:
+            actions (argparse.Action): Argument parser actions.
+        """
+        super(SortedMetavarTypeHelpFormatter, self).add_arguments(sorted(actions, key=attrgetter('option_strings')))
```
Comment (cp5555) [inline superbench/benchmarks/base.py:25]: What's the `option_strings`?

File: docs/cli.md (hunk 4 above)
Comment (cp5555) [reply docs/cli.md:29]: I think you support regex, but you don't mention here.

File: docs/cli.md (hunk 3 above)
Comment (cp5555) [reply docs/cli.md:61]: Same as https://docs.microsoft.com/en-us/cli/azure/vmss/disk?view=azure-cli-latest

File: docs/cli.md (hunk 5 above)
Comment (cp5555) [reply docs/cli.md:62]: Same as https://docs.microsoft.com/en-us/cli/azure/vmss/disk?view=azure-cli-latest

File: superbench/cli/_help.py (hunk 8 above)
Comment (cp5555) [reply superbench/cli/_help.py:83]: It's just an example

File: tests/cli/test_sb.py (hunk 6 above)
Comment (cp5555) [reply tests/cli/test_sb.py:99]: It prints directly.

**Review by cp5555 — APPROVED** [review]

File: tests/cli/test_sb.py (hunk 6 above)
Comment (guoshzhao) [inline tests/cli/test_sb.py:99]: Do we need to check the exit_code?

**Review by guoshzhao — APPROVED** [review]

File: tests/cli/test_sb.py (hunk 6 above)
Comment (abuccts) [reply tests/cli/test_sb.py:99]: by default `expect_failure` is `False` so exit code is expected to be 0, otherwise the case will fail

**Pull request merged**